File: 01/solution.js
const load = require("../util/load");

const raw = load(__dirname + "/input.txt");

/** ------------------------------------- */

const intArr = raw.split("").map((n) => parseInt(n));

function matchSum(arr, offset) {
  let sum = 0;
  for (let i = 0; i < arr.length; i++) {
    if (arr[i] === arr[(i + offset) % arr.length]) {
      sum += arr[i];
    }
  }
  return sum;
}

console.log("part 1");

const t0 = performance.now();
console.log(matchSum(intArr, 1));
const t1 = performance.now();
console.log(`(${t1 - t0}ms)`);

console.log("part 2");

const t2 = performance.now();
console.log(matchSum(intArr, intArr.length / 2));
const t3 = performance.now();
console.log(`(${t3 - t2}ms)`);

File: 01/solution.rs
use std::fs;


fn load_to_str(path: &str) -> Result<String, std::io::Error> {
    fs::read_to_string(path)
}

// ----------------------------------


fn match_sum(arr: &[i32], offset: usize) -> i32 {
    let mut sum = 0;
    for i in 0..arr.len() {
        if arr[i] == arr[(i + offset) % arr.len()] {
            sum += arr[i];
        }
    }
    sum
}

fn main () {
    let path = "input.txt";
    
    match load_to_str(path) {
        Ok(contents) => {

            let numbers: Vec<i32> = contents
                .trim()
                .chars()
                .filter_map(|c| c.to_digit(10))
                .map(|d| d as i32)
                .collect();

            println!("pt1: {}", match_sum(&numbers, 1));
            println!("pt2: {}", match_sum(&numbers, numbers.len()/2));


        }
        Err(e) => {
            eprintln!("{}", e)
        }

    }
}